attest/secureboot.go

```diff
@@ -0,0 +1,166 @@
+package attest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/go-attestation/attest/internal"
+)
+
+// SecurebootState describes the secure boot status of a machine, as determined
+// by processing its event log.
+type SecurebootState struct {
+	Enabled bool
+
+	PlatformKeys      []x509.Certificate
+	PlatformKeyHashes [][]byte
+
+	ExchangeKeys      []x509.Certificate
+	ExchangeKeyHashes [][]byte
+
+	PermittedKeys   []x509.Certificate
+	PermittedHashes [][]byte
+
+	ForbiddenKeys   []x509.Certificate
+	ForbiddenHashes [][]byte
+
+	PreSeparatorAuthority  []x509.Certificate
+	PostSeparatorAuthority []x509.Certificate
+}
+
+// ExtractSecurebootState parses a series of events to determine the
+// configuration of secure boot on a device. An error is returned if
+// the state cannot be determined, or if the event log is structured
+// in such a way that it may have been tampered post-execution of
+// platform firmware.
+func ExtractSecurebootState(events []Event) (*SecurebootState, error) {
+	// This algorithm verifies the following:
+	// - All events in PCR 7 have event types which are expected in PCR 7.
+	// - All events are parsable according to their event type.
+	// - All events have digests values corresponding to their data/event type.
+	// - No unverifiable events were present.
+	// - All variables are specified before the separator and never duplicated.
+	// - The SecureBoot variable has a value of 0 or 1.
+	// - If SecureBoot was 1 (enabled), authority events were present indicating
+	//   keys were used to perform verification.
+	// - If SecureBoot was 1 (enabled), platform + exchange + database keys
+	//   were specified.
+	// - No UEFI debugger was attached.
+
+	var (
+		out           SecurebootState
+		seenSeparator bool
+		seenAuthority bool
+		seenVars      = map[string]bool{}
+	)
+
+	for _, e := range events {
+		if e.Index != 7 {
+			continue
+		}
+
+		et, err := internal.ExtractEventType(uint32(e.Type))
+		if err != nil {
+			return nil, fmt.Errorf("unrecognised event type: %v", err)
+		}
+
+		digestVerify := e.digestEquals(e.Data)
```

Comment:
You're checking this on everything but EFIAction. Maybe

```
if et != internal.EFIAction {
    if err := e.digestEqual(e.Data); err != nil {
        // return error
    }
}
```

Reply:
That would make it cleaner, but I prefer having duplication so we get the more specific error messages in this case.